core/src/main/java/org/apache/iceberg/PartitionData.java

```diff
@@ -171,6 +169,10 @@ public PartitionData copy() {
     return new PartitionData(this);
   }
 
+  public PartitionData copyFor(StructLike partition) {
```

Comment:
An alternative to this is to create a new `StructLike` class simply backed by an array and populate it in the table scan utility.

Reply:
I find this API to be a bit awkward. I'd rather have a more straightforward two-step process to copy the key without copying values (like `emptyCopy()`) and then fill that with values. Otherwise "copy" doesn't really describe this action, which copies behavior but not data.

Reply:
I am not entirely happy too. I followed what we did in other places like StructLikeProjection.

Do we really need PartitionData here? What if we simply create a container struct backed by an array and use it? 

My worry with emptyCopy() is that the result object is in a weird state, I am not sure it is actually better than copyFor().

Reply:
We only need to a serializable projection used as a grouping key. Not more.